vault/raft.go

```diff
@@ -546,100 +567,192 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderAddr string, tlsConfig
+
+			// Create an API client to interact with the leader node
+			transport := cleanhttp.DefaultPooledTransport()
+			if leaderInfo.TLSConfig != nil {
+				transport.TLSClientConfig = leaderInfo.TLSConfig.Clone()
+				if err := http2.ConfigureTransport(transport); err != nil {
+					return errwrap.Wrapf("failed to configure TLS: {{err}}", err)
+				}
+			}
+			client := &http.Client{
+				Transport: transport,
+			}
+			config := api.DefaultConfig()
+			if config.Error != nil {
+				return errwrap.Wrapf("failed to create api client: {{err}}", config.Error)
+			}
+			config.Address = leaderInfo.LeaderAPIAddr
+			config.HttpClient = client
+			config.MaxRetries = 0
+			apiClient, err := api.NewClient(config)
+			if err != nil {
+				return errwrap.Wrapf("failed to create api client: {{err}}", err)
+			}
+
+			// Get the seal status of the leader node
+			sealStatus, err := apiClient.Sys().SealStatus()
+			if err != nil || sealStatus == nil {
+				return errwrap.Wrapf("failed to fetch seal status of leader node: {{err}}", err)
+			}
+
+			// Wait for the leader node to become active
+			if sealStatus.Sealed {
+				return errors.New("raft leader node is sealed")
+			}
+
+			// Attempt to join the leader by requesting for the bootstrap challenge
+			secret, err := apiClient.Logical().Write("sys/storage/raft/bootstrap/challenge", map[string]interface{}{
+				"server_id": raftStorage.NodeID(),
+			})
+			if err != nil {
+				return errwrap.Wrapf("error during raft bootstrap init call: {{err}}", err)
+			}
+			if secret == nil {
+				return errors.New("could not retrieve raft bootstrap package")
+			}
+
+			var sealConfig SealConfig
+			err = mapstructure.Decode(secret.Data["seal_config"], &sealConfig)
+			if err != nil {
+				return err
+			}
+
+			if sealConfig.Type != c.seal.BarrierType() {
+				return fmt.Errorf("mismatching seal types between raft leader (%s) and follower (%s)", sealConfig.Type, c.seal.BarrierType())
+			}
+
+			challengeB64, ok := secret.Data["challenge"]
+			if !ok {
+				return errors.New("error during raft bootstrap call, no challenge given")
+			}
+			challengeRaw, err := base64.StdEncoding.DecodeString(challengeB64.(string))
+			if err != nil {
+				return errwrap.Wrapf("error decoding raft bootstrap challenge: {{err}}", err)
+			}
+
+			eBlob := &physical.EncryptedBlobInfo{}
+			if err := proto.Unmarshal(challengeRaw, eBlob); err != nil {
+				return errwrap.Wrapf("error decoding raft bootstrap challenge: {{err}}", err)
+			}
+			raftInfo := &raftInformation{
+				challenge:           eBlob,
+				leaderClient:        apiClient,
+				leaderBarrierConfig: &sealConfig,
+				nonVoter:            nonVoter,
+			}
+			if c.seal.BarrierType() == seal.Shamir {
+				c.raftInfo = raftInfo
+				if err := c.seal.SetBarrierConfig(ctx, &sealConfig); err != nil {
+					return err
+				}
+				// If retry is set, continue to join after some time
+				if retry {
+					return errors.New("raft join is waiting for unseal keys to be supplied")
```

Comment:
Its a good question. The reason I left the iteration on is to ensure that all the leader info blocks are processed during the join.

But thinking about this more, I think there is an issue here. That is, if we parse and set 1 leader information in core.raftInfo and go out of this function; after supplying the unseal keys, if it turns out that that node is not the actual leader, then the process stops there and there is no way to switch to another leader information. I think we should loop out on all the leader node information while sending the answer as well.

It can be a little tricky to get this to work properly, but I think I know how to make this work. Let me get back to you on this.

Reply:
Its fixed now. We need to iterate for sure. Reason being, the `raftInfo` set in core could be of a non-leader node. If retry join is in use, the unseal process will now handover the join process back to JoinRaftCluster function instead of sending the answer itself.